src/language-js/printer-estree.js

```diff
@@ -3679,31 +3679,38 @@ function printStatementSequence(path, options, print) {
 
 function printPropertyKey(path, options, print) {
   const node = path.getNode();
+  const parent = path.getParentNode();
   const key = node.key;
 
+  const objectHasStringProp = (
+    parent.properties ||
+    parent.body ||
+    parent.members
+  ).some(
```

Comment:
This `.some()` call causes `printPropertyKey` to be `O(n^2)` in number of props because for each property being printed we look at all other properties, this may cause perf issues for very large objects. Thinking of memoizing the result with `mem`. Thoughts?

Reply:
(We also only need to compute this if `quoteProps === 'consistent'` but I left it like this during development so all the tests hit this code path.)

Reply:
You could do this check in the parent node. I’m not sure in how many contexts object keys can appear but hopefully it’s a small set.

Reply:
There are 6 calls to this function, but they're called from property levels of the AST, so I'd need to go one level up the AST again to get the parent node, difficult to figure out all the possible cases, and if we did we'd have to mutate the AST to store the info required.

Reply:
What is the `mem` thing you are talking about, I wasn't aware that we had a way to memoize computation within prettier infra. If we have it, then that would be a great fit.

If it doesn't exist, we can mutate the object for now, it's not ideal but not the end of the world either. I don't think that we're able to reuse ast right now.

Reply:
[`mem`](https://npmjs.org/package/mem) is a memoization library, we use it to cache configuration with cosmiconfig.

Thinking more though, we could probably just use a `WeakMap` where the key is the parent AST node, and the value is whether the object/class/whatever has quoted properties. We'd just create a fresh map for every top-level print call.

Reply:
Just implemented `WeakMap`. I took the simplest approach of using one module-local `WeakMap` instance, and never clearing it. If my understanding of WeakMap is correct, we should never _have_ to clear it, because as soon as the AST goes out of scope, the keys will get garbage collected and automatically removed from the map. If anyone more experienced with WeakMap thinks otherwise, let me know. 👍